crates/codegen/parser/runtime/src/cst.rs

```diff
@@ -54,63 +54,103 @@ impl Node {
         Cursor::new(self.clone(), text_offset)
     }
 
+    /// Reconstructs the original source code from the parse tree.
+    pub fn unparse(self) -> String {
+        match self {
+            Self::Rule(rule) => rule.unparse(),
+            Self::Token(token) => token.text.clone(),
+        }
+    }
+
     pub fn as_rule(&self) -> Option<&Rc<RuleNode>> {
         match self {
             Self::Rule(node) => Some(node),
             _ => None,
         }
     }
 
+    pub fn into_rule(self) -> Option<Rc<RuleNode>> {
+        match self {
+            Self::Rule(node) => Some(node),
+            _ => None,
+        }
+    }
+
     pub fn as_token(&self) -> Option<&Rc<TokenNode>> {
         match self {
             Self::Token(node) => Some(node),
             _ => None,
         }
     }
 
-    pub fn as_token_with_kind(&self, kinds: &[TokenKind]) -> Option<&Rc<TokenNode>> {
-        if let Node::Token(token_node) = self {
-            if kinds.contains(&token_node.kind) {
-                return Some(token_node);
-            }
+    pub fn into_token(self) -> Option<Rc<TokenNode>> {
+        match self {
+            Self::Token(node) => Some(node),
+            _ => None,
         }
-        return None;
+    }
+
+    pub fn as_token_with_kind(&self, kinds: &[TokenKind]) -> Option<&Rc<TokenNode>> {
+        self.as_token_matching(|token| kinds.contains(&token.kind))
     }
 
     pub fn as_token_matching<F: Fn(&Rc<TokenNode>) -> bool>(
         &self,
         predicate: F,
     ) -> Option<&Rc<TokenNode>> {
-        if let Node::Token(token_node) = self {
-            if predicate(&token_node) {
-                return Some(token_node);
-            }
+        match self {
+            Node::Token(token) if predicate(&token) => Some(token),
+            _ => None,
         }
-        return None;
     }
 
     pub fn as_rule_with_kind(&self, kinds: &[RuleKind]) -> Option<&Rc<RuleNode>> {
-        if let Node::Rule(rule_node) = self {
-            if kinds.contains(&rule_node.kind) {
-                return Some(rule_node);
-            }
-        }
-        return None;
+        self.as_rule_matching(|rule| kinds.contains(&rule.kind))
     }
 
     pub fn as_rule_matching<F: Fn(&Rc<RuleNode>) -> bool>(
         &self,
         predicate: F,
     ) -> Option<&Rc<RuleNode>> {
-        if let Node::Rule(rule_node) = self {
-            if predicate(&rule_node) {
-                return Some(rule_node);
-            }
+        match self {
+            Node::Rule(rule) if predicate(&rule) => Some(rule),
+            _ => None,
         }
-        return None;
+    }
+}
+
+impl From<Rc<RuleNode>> for Node {
+    fn from(node: Rc<RuleNode>) -> Self {
+        Self::Rule(node)
+    }
+}
+
+impl From<Rc<TokenNode>> for Node {
+    fn from(node: Rc<TokenNode>) -> Self {
+        Self::Token(node)
+    }
+}
+
+impl RuleNode {
+    /// Creates a [`Cursor`] that starts at the current node as the root and a given initial `text_offset`.
```

Comment:
should we add these methods to `TokenNode` as well?
I know the implementation is trivial, but users might not know that, and it feels a bit surprising to be able to expose this API on one kind of nodes, but not the other.

Reply:
I think that public API should discourage users from doing something "dumb" - the only use case I can see here for a cursor creation like that is a single-item iterator whereas the Cursor aims to expose much richer API.

If the user wants a single-item iterator, they can use `std::iter::once(Node::from(token_node))`. Otherwise, maybe they will mistakingly use a Token node to create a cursor and hunt the bug for some time?

I don't have very strong opinion on this as it's rather minor but I think it can cause more harm than good here.

Reply:
We could implement this for parity (as users can already call this via `Node` wrapper) but maybe add a warning in the documentation that it's something that the user almost certainly doesn't want that.

Reply:
I see. Your reasoning makes sense to me. Let's keep it as-is for now.